### PR TITLE
Kafka fixes

### LINKED ISSUE #6731: MaterializedView from Kafka Engine with Native format
Hello. 
I have the next schema:
Kakfa -> Table with Kafka Engine and Native format -> Materialized View to table -> Table with MergeTree Engine.

On Clickhouse version 19.9.2 such schema works, but on version 19.13.3 it stopped working. Got such error:
```
2019.08.29 16:02:22.127282 [ 70 ] {} <Error> void DB::StorageKafka::streamThread(): Code: 9, e.displayText() = DB::Exception: Sizes of columns doesn't match, Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x30) [0x7f6b5f0]
1. /usr/bin/clickhouse-server() [0x3944050]
2. /usr/bin/clickhouse-server(DB::SquashingTransform::add(std::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&&)+0x2c) [0x6c2e18c]
3. /usr/bin/clickhouse-server(DB::SquashingBlockOutputStream::write(DB::Block const&)+0x2ab) [0x7363f3b]
4. /usr/bin/clickhouse-server(DB::AddingDefaultBlockOutputStream::write(DB::Block const&)+0x8f) [0x72f057f]
5. /usr/bin/clickhouse-server(DB::CountingBlockOutputStream::write(DB::Block const&)+0x2c) [0x730ce3c]
6. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0x105) [0x6c2e4b5]
7. /usr/bin/clickhouse-server(DB::StorageKafka::streamToViews()+0x616) [0x729c6c6]
8. /usr/bin/clickhouse-server(DB::StorageKafka::streamThread()+0x1d7) [0x729cd97]
9. /usr/bin/clickhouse-server(DB::BackgroundSchedulePoolTaskInfo::execute()+0xf9) [0x72bd329]
10. /usr/bin/clickhouse-server(DB::BackgroundSchedulePool::threadFunction()+0x6a) [0x72bdb7a]
11. /usr/bin/clickhouse-server() [0x72bdbfa]
12. /usr/bin/clickhouse-server(ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>)+0x1a6) [0x3c760f6]
13. /usr/bin/clickhouse-server() [0xba3e1a0]
14. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3) [0x7f4868fcafa3]
15. /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f4868d734cf]
 (version 19.13.3.26 (official build))
```
Also need to say that select query from table with Kafka Engine at the same time working properly. And `insert into select` query also working properly.

Tables metadata:

`create table db.testnativekafka (Date Date, Id Int32, Name String) ENGINE = Kafka() SETTINGS kafka_broker_list = 'kafka.example.com:9092', kafka_topic_list = 'testnative', kafka_group_name = 'testnative', kafka_format = 'Native', kafka_num_consumers = 1, kafka_skip_broken_messages = 1`

`create table db.testnativetable (Date Date, Id Int32, Name String) ENGINE = MergeTree() Partition by (Date) Order By (Date)`

`create materialized view db.testnativemv to db.testnativetable (Date Date, Id Int32, Name String) as select Date, Id, Name from db.testnativekafka`

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
Fixed Native format for Kafka,  resolves #6731 #7337 #8003  
Fixed formats with headers (like CSVWithNames) which were throwing exception about eof.

Detailed description (optional):
Continuation of #7935

Those formats was tested with Kafka, and after that changes pass very basic tests: CSV CSVWithNames JSONEachRow Native RowBinary RowBinaryWithNamesAndTypes TSKV TSV TSVWithNames TSVWithNamesAndTypes

Values - is currently broken for Kafka (seems like some magic with peekable buffer), fails the test 10 msgs x 8K rows each (stops consuming after 16K rows).

~~Schema-based format was not tested yet (still TODO)~~. Tested & works (see below)

Note for backporters: **THAT PR DEPENDS ON https://github.com/ClickHouse/ClickHouse/pull/8005 !!**

### CODE REVIEW

File: dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp [hunk 1]
```diff
@@ -66,20 +65,8 @@ Block KafkaBlockInputStream::readImpl()
...
 
-    auto read_callback = [&]
-    {
-        virtual_columns[0]->insert(buffer->currentTopic());     // "topic"
-        virtual_columns[1]->insert(buffer->currentKey());       // "key"
-        virtual_columns[2]->insert(buffer->currentOffset());    // "offset"
-        virtual_columns[3]->insert(buffer->currentPartition()); // "partition"
-
-        auto timestamp = buffer->currentTimestamp();
-        if (timestamp)
```
Comment (filimonov) [inline dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp]: old producers do not write timestamp, but column still should have enough rows to be merged with main block.

**Comment by filimonov** [comment]
OK. I've tested all formats. 

'Values' format still broken, I was trying to fix that, but without success for now. Not sure if it was working before my changes (for some reason older CH version is failing all my tests).  

'Template' can't be currently be used with Kafka, as you can't pass required parameters (`DB::Exception: Unknown setting format_template_row for storage Kafka`). 

Formats tested:
CSV CSVWithNames JSONEachRow Native RowBinary RowBinaryWithNamesAndTypes TSKV TSV TSVWithNames TSVWithNamesAndTypes Values CapnProto ORC Parquet Protobuf Template

I think that PR can be merged / backported after review and getting green CI. 

I'm working now on adding also unit tests (hope after that we will be able to fix the issue with Values).

File: dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp [hunk 2]
```diff
@@ -66,20 +65,8 @@ Block KafkaBlockInputStream::readImpl()
...
-            virtual_columns[4]->insert(std::chrono::duration_cast<std::chrono::seconds>(timestamp->get_timestamp()).count()); // "timestamp"
-    };
-
     auto input_format = FormatFactory::instance().getInputFormat(
-        storage.getFormatName(), *buffer, non_virtual_header, context, max_block_size, read_callback);
```
Comment (filimonov) [inline dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp]: that callback was not called for NativeInputStream, so block with virtual columns was not filled & execption during adding virtual columns to resulting block was happened.

**Comment by filimonov** [comment]
Test used is here: https://github.com/filimonov/ch-kafka-consume-perftest/blob/master/test_formats.sh

@vzakaznikov have some plans for integarting those tests.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by KochetovNicolai** [comment]
Conflicts with 19.17
@filimonov can you please create a special pr for kafka fixes to 19.17?

**Pull request merged**